fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/JoinedRecordTypeTest.java

```diff
@@ -0,0 +1,208 @@
+
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.RecordMetaDataBuilder;
+import com.apple.foundationdb.record.TestRecordsJoinIndexProto;
+import com.apple.foundationdb.record.metadata.JoinedRecordType;
+import com.apple.foundationdb.record.metadata.JoinedRecordTypeBuilder;
+import com.apple.foundationdb.record.provider.foundationdb.query.FDBRecordStoreQueryTestBase;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.test.Tags;
+import com.google.protobuf.Message;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.ExecutionException;
+
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Tests of the {@link JoinedRecordType} class. Some of these tests may require access to an underlying FDB record
+ * store.
+ */
+@Tag(Tags.RequiresFDB)
+public class JoinedRecordTypeTest extends FDBRecordStoreQueryTestBase {
+
+    public static final String JOINED_RECORD_NAME = "JoinedRecord";
+    public static final String SIMPLE_RECORD = "simple_record";
+    public static final String OTHER_RECORD = "other_record";
+
+    @Nonnull
+    private static RecordMetaData baseMetaData(@Nonnull RecordMetaDataHook hook) {
+        RecordMetaDataBuilder metaDataBuilder = RecordMetaData.newBuilder()
+                .setRecords(TestRecordsJoinIndexProto.getDescriptor());
+        hook.apply(metaDataBuilder);
+        return metaDataBuilder.build();
+    }
+
+    @Nonnull
+    private static RecordMetaDataHook addJoinedType() {
+        return metaDataBuilder -> {
+            JoinedRecordTypeBuilder typeBuilder = metaDataBuilder.addJoinedRecordType(JOINED_RECORD_NAME);
+            typeBuilder.addConstituent(SIMPLE_RECORD, "MySimpleRecord");
+            typeBuilder.addConstituent(OTHER_RECORD, "MyOtherRecord");
+            typeBuilder.addJoin(SIMPLE_RECORD, field("other_rec_no"), OTHER_RECORD, field("rec_no"));
+        };
+    }
+
+    @Test
+    void loadSyntheticRecord() throws ExecutionException, InterruptedException {
+        final Tuple joinedPrimaryKey;
+        final FDBStoredRecord<Message> simpleRecord;
+        final FDBStoredRecord<Message> otherRecord;
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenRecordStore(context, baseMetaData(addJoinedType()));
+
+            Tuple syntheticRecordTypeKey = recordStore.getRecordMetaData()
+                    .getSyntheticRecordType(JOINED_RECORD_NAME)
+                    .getRecordTypeKeyTuple();
+            simpleRecord = recordStore.saveRecord(createSimpleRecord(100, 10));
+            otherRecord = recordStore.saveRecord(createOtherRecord(101, 11));
+            joinedPrimaryKey = Tuple.from(
```

Comment:
hmm, it would be nice if something else provided this, but I don't think we want to be improving that aspect of the synthetic record type api as part of this work.